### PR TITLE
FIX #4100 Setup the ability to overload the ShortcodeParser class and en...

### LINKED ISSUE #4100: Not able to override/overload ShortcodeParser (not a child of Object)
I'm not able to override functionality of the `ShortcodeParser` class because it's not a child of the core `Object` class like the other parsers setup under `framework/parsers`. Ideally I'd like to be able to do the following somewhere in my code (early on in execution):

```php
Object::useCustomClass('ShortcodeParser', 'CustomShortcodeParser', true);
```

I'm considering a quick PR which will tweak this class to make it a little more extensible for these sorts of situations (I'm already having to do this in my own project to address unrelated bugs).

### PR DESCRIPTION
**Pull request by patricknelson**
Setup the ability to overload the ShortcodeParser class and ensuring its methods/properties are extensible via "static" keyword. Fixes issue #4100.

### CODE REVIEW

**Comment by patricknelson** [comment]
Updated commit to change access to local properties/constants (which were changed to protected) to use `self` keyword instead of `static` keyword due to the Travis CI build errors for PHP 5.3.

File: tests/filesystem/FileTest.php [hunk 1]
```diff
@@ -12,7 +12,7 @@ class FileTest extends SapphireTest {
 	public function testLinkShortcodeHandler() {
 		$testFile = $this->objFromFixture('File', 'asdf');
 
-		$parser = new ShortcodeParser();
+		$parser = ShortcodeParser::create();
```
Comment (willmorgan) [inline tests/filesystem/FileTest.php]: I think we should keep it as it is here, as in the tests we want to be fully sure of what we're testing.

Comment (patricknelson) [inline tests/filesystem/FileTest.php]: That's a very good point. I've reverted that change.

**Comment by chillu** [comment]
Thanks Patrick! In general, we have to move those enhancements to the 3.2 release line ("3" branch here in framework), but its a low risk change so making an exception here.

**Comment by patricknelson** [comment]
Ok no problem. That being said: Do you think maybe #4140 should then be rebased upon `3` instead of `3.1` like it is now?

**Pull request merged**